gpbft/chain.go

```diff
@@ -5,16 +5,76 @@ import (
 	"encoding/binary"
 	"encoding/hex"
 	"errors"
+	"fmt"
 	"strings"
+
+	cbg "github.com/whyrusleeping/cbor-gen"
+	"golang.org/x/crypto/blake2b"
 )
 
+// TipSetKey is a
+type TipSetKey = []byte
+
+type CID = []byte
+
+var cidPrefix = []byte{0x01, 0x71, 0xA0, 0xE4, 0x02, 0x20}
+
+// Hashes the given data and returns a CBOR + blake2b-256 CID.
+func MakeCid(data []byte) []byte {
+	// TODO: Consider just using go-cid? We implicitly depend on it through cbor-gen anyways.
+	digest := blake2b.Sum256(data)
+
+	out := make([]byte, 0, 38)
+	out = append(out, cidPrefix...)
+	out = append(out, digest[:]...)
+	return out
+}
+
 // Opaque type representing a tipset.
 // This is expected to be:
 // - a canonical sequence of CIDs of block headers identifying a tipset,
 // - a commitment to the resulting power table,
 // - a commitment to additional derived values.
 // However, GossipPBFT doesn't need to know anything about that structure.
-type TipSet = []byte
+type TipSet struct {
+	Epoch       int64
+	TipSet      TipSetKey
```

Comment:
```suggestion
	Key?      TipSetKey
```

This will probably often show up as `tipset.TipSet`